types/wlr_cursor.c

```diff
@@ -193,53 +198,81 @@ static void cursor_warp_unchecked(struct wlr_cursor *cur,
 }
 
 /**
- * Get the most specific mapping box for the device in this order:
+ * Gets the intersection of all mapping boxes for this
+ * input device.
  *
- * 1. device geometry mapping
- * 2. device output mapping
- * 3. cursor geometry mapping
- * 4. cursor output mapping
+ * If no box is found, false is returned, otherwise true.
+ * The box is written to the out parameter.
  *
  * Absolute movement for touch and pen devices will be relative to this box and
  * pointer movement will be constrained to this box.
  *
  * If none of these are set, returns NULL and absolute movement should be
  * relative to the extents of the layout.
  */
-static struct wlr_box *get_mapping(struct wlr_cursor *cur,
-		struct wlr_input_device *dev) {
+static bool get_mapping(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, struct wlr_box *out) {
 	assert(cur->state->layout);
+	*out = WLR_BOX_MAX;
+
+	bool valid = false;
+
 	struct wlr_cursor_device *c_device = get_cursor_device(cur, dev);
 
 	if (c_device) {
 		if (c_device->mapped_box) {
-			return c_device->mapped_box;
+			wlr_box_intersection(out, c_device->mapped_box, out);
+			valid = true;
 		}
 		if (c_device->mapped_output) {
-			return wlr_output_layout_get_box(cur->state->layout,
+			struct wlr_box *layout_box = wlr_output_layout_get_box(cur->state->layout,
 				c_device->mapped_output);
+			wlr_box_intersection(out, layout_box, out);
+			valid = true;
 		}
 	}
 
 	if (cur->state->mapped_box) {
-		return cur->state->mapped_box;
+		wlr_box_intersection(out, cur->state->mapped_box, out);
+		valid = true;
 	}
 	if (cur->state->mapped_output) {
-		return wlr_output_layout_get_box(cur->state->layout,
+		struct wlr_box *layout_box = wlr_output_layout_get_box(cur->state->layout,
 			cur->state->mapped_output);
+		wlr_box_intersection(out, layout_box, out);
+		valid = true;
 	}
 
-	return NULL;
+	return valid;
+}
+
+/**
+ * Gets the intersection of all confines for this
+ * input device.
+ * Confines are a superset of mappings.
+ *
+ * This function wraps get_mapping.
+ */
+static bool get_confine(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, struct wlr_box *out) {
+	bool valid = get_mapping(cur, dev, out);
+
+	if (cur->state->confined_box) {
+		wlr_box_intersection(out, cur->state->confined_box, out);
```

Comment:
what if there's no intersection?

Reply:
That is a good question, and I can't even see how wlr_box_intersection handles non-intersecting boxes.
However, in the case of pointer constraints, this should be impossible, since a pointer constraint can only be active if a cursor is on it, inside the mapping. If the mapping later changes though, then I'm not entirely sure.

Reply:
I think it returns false if there's no intersection. Maybe you set `valid` to false if there is no intersection and it will be handled later on?